[DOC-743] YSQL: Alter table rewrite docs (#27004)
yugabyte/yugabyte-db: docs/content/stable/api/ysql/the-sql-language/statements/ddl_alter_table.md

Add the specified column with the specified data type and constraint.

##### Table rewrites

> ADD COLUMN … DEFAULT statements require a [table rewrite](#alter-table-operations-that-involve-a-table-rewrite) when the default value uses *volatile* functions. [Volatile functions](https://www.postgresql.org/docs/current/xfunc-volatility.html#XFUNC-VOLATILITY) can return different results for different rows, so a table rewrite is required to fill in values for existing rows. For non-volatile functions, no table rewrite is required.

Reviewer: nit: i think volatile expressions is more accurate (similarly non-volatile expressions)

Reviewer: when the default value is a volatile expression?

Author: sounds good, will do.

Add the specified constraint to the table.

##### Table rewrites

Adding a `PRIMARY KEY` constraint results in a full table rewrite of the main table and all associated indexes, which can be a potentially expensive operation. For more details about [table rewrites, see this section](#alter-table-operations-that-involve-a-table-rewrite).

> The reason for the table rewrite is the clustered storage by primary key that YugabyteDB uses to store rows and indexes. Tables without a `PRIMARY KEY` have a hidden one underneath and rows are stored clustered on it. These rows need to be rewritten to use the newly added primary key column. 

Reviewer 2: ```suggestion
The table rewrite is needed because of how YugabyteDB stores rows and indexes. In YugabyteDB, data is distributed based on the primary key; when a table does not have an explicit primary key assigned, YugabyteDB automatically creates an internal row ID to use as the table's primary key. As a result, these rows need to be rewritten to use the newly added primary key column. For more information, refer to [Primary keys](../../../../../develop/data-modeling/primary-keys-ysql). 
```

Author: Thanks this is much more clear!

#### ALTER [ COLUMN ] *column_name* [ SET DATA ] TYPE *data_type* [ COLLATE *collation* ] [ USING *expression* ]

Change the type of an existing column. The following semantics apply:
- If the optional `COLLATE` clause is not specified, the default collation for the new column type will be used.
- If the optional `USING` clause is not provided, the default conversion for the new column value will be the same as an assignment cast from the old type to the new type.
- A `USING` clause must be included when there is no implicit assignment cast available from the old type to the new type.
- Alter type is not supported for partitioned tables. See [#16980](https://github.com/yugabyte/yugabyte-db/issues/16980).
- Alter type is not supported for tables with rules (limitation inherited from PostgreSQL).
- Alter type is not supported for tables with CDC streams, or xCluster replication when it requires data on disk to change. See [#16625](https://github.com/yugabyte/yugabyte-db/issues/16625).

##### Table rewrites

> Altering a column’s type requires a [full table rewrite](#alter-table-operations-that-involve-a-table-rewrite) of the table and any indexes that contain this column when the underlying storage format changes or if the data changes.

Reviewer 2: ```suggestion
Altering a column's type requires a [full table rewrite](#alter-table-operations-that-involve-a-table-rewrite) of the table, and any indexes that contain this column when the underlying storage format changes or if the data changes.
```

> The following type changes are common cases where we require a table rewrite:

Reviewer 2: ```suggestion
The following type changes commonly require a table rewrite:
```

> |     From         |  To               | Reason for table rewrite                                       |

Reviewer 2: Maybe the Primary key page I linked above should have a link to this. How useful is this info? Is it buried here and should rather/additionally go in Architecture or develop or somewhere?

| From              |  To                   | Notes                  |
| ------------ | ------------------ | ------------------------------------------------------ |
| VARCHAR(n)   | VARCHAR(m) (m > n) | Length increase is compatible.                         |
| VARCHAR(n)   | TEXT               | Always compatible.                                     |
| SERIAL       | INTEGER            | Underlying type is INTEGER; usually OK.                |
| NUMERIC(p,s) | NUMERIC(p2,s2)     | If new precision is larger and scale remains the same. |
| CHAR(n)      | CHAR(m) (m > n)    | PG stores it as padded TEXT, so often fine.            |
| Domain types | Their base type    | Compatible, unless additional constraints exist.       |

Altering a column with a (non-trivial) USING clause always requires a rewrite.

The table rewrite operation preserves split properties for hash-partitioned tables and hash-partitioned secondary indexes. For range-partitioned tables (and secondary indexes), split properties are only preserved if the altered column is not part of the table's (or secondary index's) range key.

> Examples of ALTER TYPE that cause a table rewrite

Reviewer 2: ```suggestion
For example, the following ALTER TYPE statements would cause a table rewrite:
```

- ALTER TABLE foo
    ALTER COLUMN foo_timestamp TYPE timestamp with time zone
    USING
        timestamp with time zone 'epoch' + foo_timestamp * interval '1 second';
- ALTER TABLE t ALTER COLUMN t_num1 TYPE NUMERIC(9,5) -- from NUMERIC(6,1);
- ALTER TABLE test ALTER COLUMN a SET DATA TYPE BIGINT USING a::BIGINT; -- from INT

> Examples of ALTER TYPE that do not cause a table rewrite

Reviewer 2: ```suggestion
The following ALTER TYPE statement does not cause a table rewrite:
```

- `RESTRICT` — Remove only the specified constraint.
- `CASCADE` — Remove the specified constraint and any dependent objects.

##### Table rewrites

> Dropping the `PRIMARY KEY` constraint results in a full table rewrite and full rewrite of all indexes associated with the table, which is a potentially expensive operation. More details and common limitations of table rewrites  [are described in this section](#alter-table-operations-that-involve-a-table-rewrite).

Reviewer 2: ```suggestion
Dropping the `PRIMARY KEY` constraint results in a full table rewrite and full rewrite of all indexes associated with the table, which is a potentially expensive operation. For more details and common limitations of table rewrites, refer to [Alter table operations that involve a table rewrite](#alter-table-operations-that-involve-a-table-rewrite).
```

Constraints marked as `INITIALLY DEFERRED` will be checked at the end of the transaction.

## Alter table operations that involve a table rewrite

Most ALTER TABLE statements only involve a schema modification and complete quickly. However, certain specific ALTER TABLE statements require a new copy of the underlying table (and associated index tables, in some cases) to be made and can potentially take a long time, depending on the sizes of the tables and indexes involved. This is typically referred to as a "table rewrite". This behavior is [similar to PostgreSQL](https://www.crunchydata.com/blog/when-does-alter-table-require-a-rewrite), though the exact scenarios when a rewrite is triggered may differ between PostgreSQL and YugabyteDB.

> It is not safe to execute concurrent DML on the table during a table rewrite because the results of any concurrent DML are not guaranteed to be reflected in the copy of the table being made. This behavior is also similar to PostgreSQL, where a table lock typically excludes concurrent DML on the table during the rewrite.

Reviewer 3: `This behavior is also similar to PostgreSQL, where a table lock typically excludes concurrent DML on the table during the rewrite.`

This indicates that PG has the consistency issues as well. May be slight bit rewrite. 

`PostgreSQL also has the same behaviour and safeguards the TableRewrite operation by  preventing the concurrent write operation by using the table lock.
`

> When such expensive rewrites have to be performed, it is recommended to combine them into a single ALTER TABLE as shown below to avoid multiple expensive rewrites.

Reviewer 2: ```suggestion
If you need to perform one of these expensive rewrites, it is recommended to combine them into a single ALTER TABLE statement to avoid multiple expensive rewrites. For example:
```

```
ALTER TABLE t ADD COLUMN c6 UUID DEFAULT gen_random_uuid(), ALTER COLUMN c8 TYPE TEXT
```

> The following alter table operations involve making a full copy of the underlying table (and possibly associated index tables).

Reviewer 2: ```suggestion
The following ALTER TABLE operations involve making a full copy of the underlying table (and possibly associated index tables):

```
